File: anchor/programs/fogopulse/src/state/epoch.rs
use anchor_lang::prelude::*;

/// Epoch state machine - tracks the lifecycle of a trading epoch
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace, Default, Debug)]
pub enum EpochState {
    /// Trading allowed
    #[default]
    Open,
    /// In freeze window, no trades allowed
    Frozen,
    /// Settlement in progress
    Settling,
    /// Outcome determined, payouts available
    Settled,
    /// Oracle failed, all positions refunded
    Refunded,
}

impl EpochState {
    /// Convert to u8 for Pool.active_epoch_state cache field
    /// 0 = None (no active epoch), 1 = Open, 2 = Frozen, etc.
    pub const fn as_pool_cache_u8(&self) -> u8 {
        match self {
            EpochState::Open => 1,
            EpochState::Frozen => 2,
            EpochState::Settling => 3,
            EpochState::Settled => 4,
            EpochState::Refunded => 5,
        }
    }
}

/// Final outcome of an epoch after settlement
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace, Default)]
pub enum Outcome {
    /// Settlement price > Start price
    #[default]
    Up,
    /// Settlement price < Start price
    Down,
    /// Confidence bands overlap or prices tied - all refunded
    Refunded,
}

/// Reason for epoch refund (for event logging)
#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace)]
pub enum RefundReason {
    /// Confidence bands overlap - outcome too uncertain
    ConfidenceOverlap,
    /// Settlement price exactly equals start price
    Tie,
}

/// Epoch account - represents a time-bounded trading period within a pool
#[account]
#[derive(InitSpace)]
pub struct Epoch {
    /// Parent pool reference
    pub pool: Pubkey,

    /// Sequential identifier within pool (0, 1, 2, ...)
    pub epoch_id: u64,

    /// Current epoch state
    pub state: EpochState,

    /// Unix timestamp when epoch begins
    pub start_time: i64,

    /// Unix timestamp when epoch ends
    pub end_time: i64,

    /// When trading stops (end_time - freeze_window_seconds)
    pub freeze_time: i64,

    /// Oracle price at epoch creation
    pub start_price: u64,

    /// Oracle confidence at epoch creation
    pub start_confidence: u64,

    /// Oracle publish timestamp at epoch creation
    pub start_publish_time: i64,

    /// Oracle price at settlement (None until settled)
    pub settlement_price: Option<u64>,

    /// Oracle confidence at settlement
    pub settlement_confidence: Option<u64>,

    /// Oracle publish timestamp at settlement
    pub settlement_publish_time: Option<i64>,

    /// Final outcome (Up, Down, or Refunded)
    pub outcome: Option<Outcome>,

    /// PDA bump seed
    pub bump: u8,
}

File: web/src/types/epoch.ts
import type { PublicKey } from '@solana/web3.js'

/**
 * Epoch state machine - matches on-chain EpochState enum from anchor program
 */
export enum EpochState {
  /** Trading allowed */
  Open = 'Open',
  /** In freeze window, no trades allowed */
  Frozen = 'Frozen',
  /** Settlement in progress */
  Settling = 'Settling',
  /** Outcome determined, payouts available */
  Settled = 'Settled',
  /** Oracle failed, all positions refunded */
  Refunded = 'Refunded',
}

/**
 * Map the on-chain u8 representation to EpochState enum
 * From Pool.active_epoch_state cache field:
 * 0 = None (no active epoch), 1 = Open, 2 = Frozen, 3 = Settling, 4 = Settled, 5 = Refunded
 */
export function epochStateFromU8(value: number): EpochState | null {
  switch (value) {
    case 0:
      return null
    case 1:
      return EpochState.Open
    case 2:
      return EpochState.Frozen
    case 3:
      return EpochState.Settling
    case 4:
      return EpochState.Settled
    case 5:
      return EpochState.Refunded
    default:
      return null
  }
}

/**
 * Final outcome of an epoch after settlement
 */
export enum Outcome {
  /** Settlement price > Start price */
  Up = 'Up',
  /** Settlement price < Start price */
  Down = 'Down',
  /** Exact price tie - all refunded */
  Refunded = 'Refunded',
}

/**
 * Epoch data interface - matches on-chain Epoch account structure
 */
export interface EpochData {
  /** Parent pool pubkey reference */
  pool: PublicKey
  /** Sequential identifier within pool (0, 1, 2, ...) */
  epochId: bigint
  /** Current epoch state */
  state: EpochState
  /** Unix timestamp when epoch begins (seconds) */
  startTime: number
  /** Unix timestamp when epoch ends (seconds) */
  endTime: number
  /** When trading stops (end_time - freeze_window_seconds) (seconds) */
  freezeTime: number
  /** Oracle price at epoch creation (scaled u64) */
  startPrice: bigint
  /** Oracle confidence at epoch creation (scaled u64) */
  startConfidence: bigint
  /** Oracle publish timestamp at epoch creation (seconds) */
  startPublishTime: number
  /** Oracle price at settlement (None until settled) */
  settlementPrice: bigint | null
  /** Oracle confidence at settlement */
  settlementConfidence: bigint | null
  /** Oracle publish timestamp at settlement */
  settlementPublishTime: number | null
  /** Final outcome (Up, Down, or Refunded) */
  outcome: Outcome | null
  /** YES side total at settlement (before rebalance) - for payout calculation */
  yesTotalAtSettlement: bigint | null
  /** NO side total at settlement (before rebalance) - for payout calculation */
  noTotalAtSettlement: bigint | null
  /** PDA bump seed */
  bump: number
}

/**
 * Processed epoch state for UI consumption
 */
export interface EpochUIState {
  /** The raw epoch data, null if no active epoch */
  epoch: EpochData | null
  /** Time remaining in seconds until epoch end */
  timeRemaining: number
  /** Whether the epoch is in frozen state (trading stopped) */
  isFrozen: boolean
  /** Whether the epoch is currently settling */
  isSettling: boolean
  /** Whether the epoch has been settled */
  isSettled: boolean
  /** Human-readable start price (converted from scaled u64) */
  startPriceDisplay: number | null
  /** Pyth exponent for price formatting (typically -8) */
  priceExponent: number
}

/**
 * Status when no epoch is active
 */
export type NoEpochStatus = 'no-pool' | 'no-epoch' | 'next-epoch-soon'

/**
 * Parse on-chain epoch state from Anchor enum format (e.g., { open: {} })
 */
export function parseEpochState(state: unknown): EpochState {
  if (!state || typeof state !== 'object') return EpochState.Open
  const keys = Object.keys(state)
  if (keys.length === 0) return EpochState.Open
  const variant = keys[0]
  switch (variant) {
    case 'open':
      return EpochState.Open
    case 'frozen':
      return EpochState.Frozen
    case 'settling':
      return EpochState.Settling
    case 'settled':
      return EpochState.Settled
    case 'refunded':
      return EpochState.Refunded
    default:
      return EpochState.Open
  }
}

/**
 * Parse outcome from Anchor enum format (e.g., { up: {} })
 */
export function parseOutcome(outcome: unknown): Outcome | null {
  if (!outcome || typeof outcome !== 'object') return null
  const keys = Object.keys(outcome)
  if (keys.length === 0) return null
  const variant = keys[0]
  switch (variant) {
    case 'up':
      return Outcome.Up
    case 'down':
      return Outcome.Down
    case 'refunded':
      return Outcome.Refunded
    default:
      return null
  }
}
